src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,1297 @@
+        res.ctx = (<fmpz_mod_poly>left).ctx
+        check = fmpz_mod_poly_divides(
+            res.val, (<fmpz_mod_poly>left).val, (<fmpz_mod_poly>right).val, res.ctx.mod.val
+        )
+        if check == 0:
+            raise ValueError(
+                f"{right} does not divide {left}"
+            )
+
+        return res
+
+    def __truediv__(s, t):
+        return fmpz_mod_poly._div_(s, t)
+
+    def __rtruediv__(s, t):
+        return fmpz_mod_poly._div_(t, s)
+
+    @staticmethod
+    def _floordiv_(left, right):
+        cdef fmpz_mod_poly res
+
+        # Case when left and right are already fmpz_mod_poly
+        if typecheck(left, fmpz_mod_poly) and typecheck(right, fmpz_mod_poly):
+            if not (<fmpz_mod_poly>left).ctx == (<fmpz_mod_poly>right).ctx:
+                raise ValueError("moduli must match")
+
+        # Case when right is not fmpz_mod_poly, try to convert to fmpz
+        elif typecheck(left, fmpz_mod_poly):
+            right = (<fmpz_mod_poly>left).ctx.any_as_fmpz_mod_poly(right)
+            if right is NotImplemented:
+                return NotImplemented
+
+        # Case when left is not fmpz_mod_poly, try to convert to fmpz
+        else:
+            left = (<fmpz_mod_poly>right).ctx.any_as_fmpz_mod_poly(left)
+            if left is NotImplemented:
+                return NotImplemented
+
+        if not right.leading_coefficient().is_unit():
+            raise ValueError(f"The leading term of {right} must be a unit modulo N")
+
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = (<fmpz_mod_poly>left).ctx
+        fmpz_mod_poly_div(
+            res.val, (<fmpz_mod_poly>left).val, (<fmpz_mod_poly>right).val, res.ctx.mod.val
+        )
+        return res 
+
+    def __floordiv__(self, other):
+        return fmpz_mod_poly._floordiv_(self, other)
+
+    def __rfloordiv__(self, other):
+        return fmpz_mod_poly._floordiv_(other, self)
+
+    def __pow__(self, e):
+        cdef fmpz_mod_poly res
+        if e < 0:
+            raise ValueError("Exponent must be non-negative")
+
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = self.ctx
+        fmpz_mod_poly_pow(
+            res.val, self.val, (<ulong>e), self.ctx.mod.val
+        )
+        return res
+
+    def shift(self, slong n):
+        """
+        Returns ``self`` shifted by ``n`` coefficients. If ``n`` is positive,
+        zero coefficients are inserted, when ``n`` is negative, if ``n`` is 
+        greater than or equal to the length of ``self``, the zero polynomial
+        is returned.
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> f = R([1,2,3])
+            >>> f.shift(0)
+            3*x^2 + 2*x + 1
+            >>> f.shift(1)
+            3*x^3 + 2*x^2 + x
+            >>> f.shift(4)
+            3*x^6 + 2*x^5 + x^4
+            >>> f.shift(-1)
+            3*x + 2
+            >>> f.shift(-4)
+            0
+
+        """
+        if n == 0:
+            return self
```

Comment:
In general I wonder how we want to handle things like this. Usually in Python if an object is mutable then you want to promise whether any particular function will return a copy or not. With immutable objects it is safe to always return self. Otherwise if mutability is exposed then there needs to be a decision everywhere about whether copies are to be made or not.

Reply:
I figured returning self here was OK and it's faster, but I suppose another option is to simply create a new polynomial and set the value to get a distinct object. I don't have a big opinion

Reply:
If polynomials are mutable then there needs to be a way to copy them. Most mutable types in Python have a `.copy()` method for this.

It should be reasonably fast. Returning a copy should be considered correct behaviour unless there is some documentation describing when copies might not be made.

Reply:
I guess we also need to decide whether to implement things like `a += b` if there is a need to allow mutability.

It is not clear to me that there is a strong benefit in allowing mutability in general for poly although it is most likely beneficial for matrix types. At the Python level you probably make these polynomials from a list anyway and you can just mutate the list if you are doing that.